conbench/api/runs.py

```diff
@@ -1,26 +1,64 @@
+import datetime
+from typing import Dict, List, Optional, Tuple
+
 import flask as f
 import flask_login
 
 from ..api import rule
 from ..api._docs import spec
 from ..api._endpoint import ApiEndpoint, maybe_login_required
-from ..entities._entity import EntityExists, NotFound
 from ..entities.benchmark_result import BenchmarkResult
 from ..entities.commit import Commit
-from ..entities.run import Run, RunFacadeSchema, RunSerializer
+from ..entities.run import RunFacadeSchema, RunSerializer
+
+
+def get_all_run_info(
+    min_time: datetime.datetime,
+    max_time: datetime.datetime,
+    commit_hashes: Optional[List[str]] = None,
+) -> List[Tuple[BenchmarkResult, int, bool]]:
+    """Get info about each run in a time range.
+
+    Query the database for all benchmark results (optionally filtered to certain commit
+    hashes) in a time range, and return a list of tuples, each corresponding to one
+    run_id. The tuples contain:
+
+    - the earliest BenchmarkResult for that run (by its timestamp)
+    - the number of benchmark results seen for that run
+    - whether any benchmark results are "failed" in that run
+
+    This powers the "list runs" API endpoint and the "recent runs" list on the home page.
+    """
+    run_info: Dict[str, Tuple[BenchmarkResult, int, bool]] = {}  # keyed by run_id
+
+    joins = []
+    filters = [
+        BenchmarkResult.timestamp >= min_time,
+        BenchmarkResult.timestamp <= max_time,
+    ]
+    if commit_hashes:
+        joins.append(Commit)
+        filters.append(Commit.sha.in_(commit_hashes))
+
+    for result in BenchmarkResult.search(filters=filters, joins=joins):
+        run_id = result.run_id
+        if run_id not in run_info:
+            run_info[run_id] = (result, 1, result.is_failed)
+        else:
+            old_result, old_count, old_failed = run_info[run_id]
+            run_info[run_id] = (
+                result if result.timestamp < old_result.timestamp else old_result,
+                old_count + 1,
+                old_failed or result.is_failed,
+            )
```

Comment:
You could write this with `functools.reduce()`. Not sure it's better, really, but for me as someone who's written a lot of functional code it's a bit quicker to mentally parse